### PR TITLE
API Key Management Refactoring & Integration Testing (Issue #305)

### LINKED ISSUE #305: Write SDK integration tests for API Key
We need to add integration tests in the Python SDK to validate API key behavior across different permission levels. These tests should ensure that read/write operations are correctly enforced based on the permissions associated with each API key.

Additionally, the test script (`scripts/tests.sh`) should be extended to support running the server with API key authentication enabled and passing the generated key into the test suite.

#### SDK (`mosaico-sdk-py`)

* Create a new file:
  ```
  src/testing/integration/test_auth.py
  ```
* Use an API key with **"manage"** permissions:
  * This key will be generated by the backend.
  * It should be passed via pytest CLI using `--api-key`:
* Inside the tests:
  * Use the manage-level API key to generate additional API keys with varying permission levels.
  * Write test cases to validate:
    * Correct behavior of **read operations** based on permissions.
    * Correct behavior of **write operations** based on permissions.
    * Proper restriction/authorization failures when permissions are insufficient.

#### Test Script (`scripts/tests.sh`)

* Add a new testing block that:
  * Launches the server with API key authentication enabled:
  * Generates an API key with **manage** permissions.
  * Passes the generated API key to pytest via:
    ```
    pytest <args> --api-key <generated_key>
    ```

### Expected Outcome

* Comprehensive integration tests validating API key permission handling.
* Automated test flow that:
  * Spins up the server with API key support.
  * Generates and injects a valid API key into the test suite.
* Clear validation of authorization logic across different permission levels.

### Notes

* Ensure tests are isolated and do not interfere with each other (e.g., clean up created API keys if necessary).
* Consider edge cases such as:
  * Invalid API keys
  * Expired or revoked keys (if applicable)
  * Missing API key in requests

### PR DESCRIPTION
**Pull request by fdicorato**
This PR refactors the API key management system (SDK) to improve error handling and API consistency, while establishing comprehensive integration tests for API key operations. 

#### API Key Module Improvements
- **Fingerprint parsing**: Extracted API key fingerprint parsing into dedicated `_get_fingerprint()` helper function
- **API consistency**: Changed `api_key_create()` to accept singular `permission` parameter instead of list for cleaner API
- **`APIKeyStatus` class**: Removed `api_key_fingerprint` field (use fingerprint for status queries directly)
- **Error handling**: Updated all error handlers to raise exceptions instead of returning defaults, eliminating silent failures

#### Client & Middleware Enhancements
- **`MosaicoClient`**: Use dict `.pop()` for safer cache removal operations
- **Error handling**: Improved base_session_writer error handling and documentation
- **Middleware**: Enhanced error responses in communication layer

#### Test Improvements
- **Test infrastructure**: Reorganized fixtures from root `conftest.py` into integration-specific `src/testing/integration/conftest.py`
- **Fixture clarity**: Renamed key fixtures for better semantics:
  - `_client` → `mosaico_client`
  - `_make_sequence_data_stream` → `synthetic_sequence_data_stream`
- **API Key testing**: Added `--full-stack-api-key` option with automatic API key generation for E2E testing
- **Comprehensive coverage**: Added `test_auth` integration tests for validating API key-based operation authentication
- **Test script**: Updated tests.sh with support for new testing modes

#### Documentation
- Improved docstrings with specific exception types across communication layer

### Test Coverage
All 23 integration tests updated to use new fixture API:
- `test_endpoints.py`, `test_error_policy.py`, `test_failures.py`
- `test_handlers.py`, `test_list_flights.py`, `test_query_*.py`
- `test_sequence_update.py`, `test_stream*.py`, `test_synch_transformer.py`, `test_to_pandas.py`

### Impact
- **API Design**: Cleaner, more consistent API for API key management
- **Reliability**: Better error handling with explicit exceptions instead of silent failures
- **Testing**: Robust integration testing for authentication workflows
- **Maintainability**: Better organized test fixtures and improved documentation

Closes #305

### CODE REVIEW

File: mosaico-sdk-py/src/mosaicolabs/handlers/base_session_writer.py
Comment (LuigiTer) [inline mosaico-sdk-py/src/mosaicolabs/handlers/base_session_writer.py:225]: Maybe `exc_type is None` is useless as check because at start of the function you declare `error_in_block = exc_type is not None` and here you are in the if block `if error_in_block:`

File: mosaico-sdk-py/src/testing/integration/test_auth.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+        _get_fingerprint("msco_58qb7dssul32r1bewpziy3rfjuewd0a3")
+
+    with pytest.raises(ValueError, match="not alnum"):
+        # Payload not alpha AND num
+        _get_fingerprint("msco_123457678_abcd1234")
+
+    with pytest.raises(ValueError, match="fingerprint"):
+        # Fingerprint less than 8 char
+        _get_fingerprint("msco_58qb7dssul_c5505")
+
```
Comment (LuigiTer) [inline mosaico-sdk-py/src/testing/integration/test_auth.py:183]: Add test case for:
* payload all letters
* payload all numbers
* payload contains special characters like `!`

File: mosaico-sdk-py/src/testing/integration/test_auth.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    host,
+    port,
+    api_keys_list: List[Tuple],
+    mosaico_client: MosaicoClient,
+    inject_synthetic_sequence,
+):
+    if not with_auth:
+        pytest.skip("Tests run without '--api-key'")
+
+    # extract a Read-Only API Key among the one created
```
Comment (LuigiTer) [inline mosaico-sdk-py/src/testing/integration/test_auth.py]: Maybe not `Read-Only`

File: mosaico-sdk-py/src/mosaicolabs/handlers/base_session_writer.py
Comment (fdicorato) [reply mosaico-sdk-py/src/mosaicolabs/handlers/base_session_writer.py:225]: The point here is:
* `exc_type` is not None only if we are exiting the context due to an exception inside the `with` block
* the `__exit__` function returns None, so if an exception happens inside the context, the function will not suppress exceptions.
* `error_in_block` is set in line 193 also, i.e. if the program is exiting the `with` context without any error, but some errors happen inside the __exit__ method, while finalizing the session.

The logic at line 225 is: if you are not exiting the context due to an error (`exc_type is None`) but some errors happen during the closing logics, log and re-raise (do not suppress).
